Add View Filters

GCRA101 opened the pull request:
### Issues addressed by this PR
<!-- Add reference(s) to issue(s) solved by this PR. Please use keyword Fixes/Closes as per https://help.github.com/articles/closing-issues-using-keywords/ -->

Closes #1501 


![Test GH Script - All Algorithms](https://github.com/user-attachments/assets/19e16d5f-1584-4b1b-acd9-056cc828a5b4)




![Test Revit Model - 3D View Filters Table](https://github.com/user-attachments/assets/86bfbed5-13a8-4918-96f1-20444aa52c26)





It is now possible to push, pull and update View Filters in Revit via the BHoM and assign to them OverrideGraphicSettings in selected Views.


### Test files
#### Grasshopper File
[https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/Revit_Toolkit/%231502-AddViewFilters/[TestScript.gh](https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/Revit_Toolkit/%231502-AddViewFilters/TestScript.gh?csf=1&web=1&e=l3Fafx)?csf=1&web=1&e=l3Fafx](url)
#### Revit Files
[https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/Revit_Toolkit/%231502-AddViewFilters/[TestRevitfile_PushModel.rvt](https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/Revit_Toolkit/%231502-AddViewFilters/TestRevitfile_PushModel.rvt?csf=1&web=1&e=O8L8fo)?csf=1&web=1&e=O8L8fo](url)
[https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/Revit_Toolkit/%231502-AddViewFilters/[TestRevitfile_PullModel.rvt](https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/Revit_Toolkit/%231502-AddViewFilters/TestRevitfile_PullModel.rvt?csf=1&web=1&e=ge2KZI)?csf=1&web=1&e=ge2KZI](url)
[https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/Revit_Toolkit/%231502-AddViewFilters/[TestRevitfile_UpdateModel.rvt](https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/Revit_Toolkit/%231502-AddViewFilters/TestRevitfile_UpdateModel.rvt?csf=1&web=1&e=yj4ExY)?csf=1&web=1&e=yj4ExY](url)


### Changelog

- FilterRules classes mirroring the corresponding Revit API ones (all within a new namespace FilterRules)
- New Elements classes View, ViewFilter and OverrideGraphicSettings
- Conversion methods in the Convert partial class allowing to convert BHoM ViewFilter objects into/from Revit ParameterFilterElement

pawelbaran committed:
draft oM classes added

pawelbaran committed:
draft oM classes added

pawelbaran committed:
draft support for pull of views and view filters added

pawelbaran committed:
draft support for pull of views and view filters added

pawelbaran committed:
draft support for views and view filters added

pawelbaran committed:
draft support for views and view filters added

GCRA101 committed:
Merge branch 'Revit_Toolkit-#1444-FilterPrototype' of https://github.com/BHoM/Revit_Toolkit into Revit_Toolkit-#1444-FilterPrototype

GCRA101 committed:
Add bespoke classes defining graphics settings for new view filters
	- Enum LinePattern
	- Enum SolidPattern
	- Class OverrideGraphicSettings
	- Class FilterRule

GCRA101 committed:
Add bespoke classes defining graphics settings for new view filters
	- Enum LinePattern
	- Enum SolidPattern
	- Class OverrideGraphicSettings
	- Class FilterRule

GCRA101 committed:
Rename Class FilterWithOverrides to ViewFilterWithOverrides

GCRA101 committed:
Rename Class FilterWithOverrides to ViewFilterWithOverrides

GCRA101 committed:
Add FilterRuleType Enumeration
	- Enumeration Class that works as an attribute of the FilterRule Class, allowing to
	  avoid the use of several abstract and concrete classes for each type of rule type.

GCRA101 committed:
Add FilterRuleType Enumeration
	- Enumeration Class that works as an attribute of the FilterRule Class, allowing to
	  avoid the use of several abstract and concrete classes for each type of rule type.

GCRA101 committed:
Move classes FillPattern, FilterRule and LinePattern from Elements to Views namespace

GCRA101 committed:
Move classes FillPattern, FilterRule and LinePattern from Elements to Views namespace

GCRA101 committed:
Implement Update of View based on filters and overrides

GCRA101 committed:
Implement Update of View based on filters and overrides

GCRA101 committed:
Add minor updates to main classes in Elements namespace

GCRA101 committed:
Add minor updates to main classes in Elements namespace

GCRA101 committed:
Implement Conversion of BHoM classes to Revit

GCRA101 committed:
Implement Conversion of BHoM classes to Revit

GCRA101 committed:
Add minor updates to BHoM ViewFilterOverrides Class

GCRA101 committed:
Add minor updates to BHoM ViewFilterOverrides Class

GCRA101 committed:
Update Query/BHoMType.cs class for View and ParameterFilterElement

GCRA101 committed:
Update Query/BHoMType.cs class for View and ParameterFilterElement

GCRA101 committed:
Minor updates to main new classes for View and filtering

GCRA101 committed:
Minor updates to main new classes for View and filtering

GCRA101 committed:
Merge branch 'Revit_Toolkit-#1444-FilterPrototype' of https://github.com/BHoM/Revit_Toolkit into Revit_Toolkit-#1444-FilterPrototype

GCRA101 committed:
Add FilterRuleEvaluatorType Enumeration

GCRA101 committed:
Add FilterRuleEvaluatorType Enumeration

GCRA101 committed:
Add FilterValueRule and modify FilterRuleType

To simplify the mirroring representation of the Revit API classes within the BHoM, it's necessary to add a concrete classe, subclass of FilterRule, that
corresponds to the FilterValueRule from Revit API.
Whatever sub-class type as well as Evaluator class type will be represented in BHoM via enumeration classes assigned to the RevitFilterRule and RevitFilterValueRule
as attributes.

GCRA101 committed:
Add FilterValueRule and modify FilterRuleType

To simplify the mirroring representation of the Revit API classes within the BHoM, it's necessary to add a concrete classe, subclass of FilterRule, that
corresponds to the FilterValueRule from Revit API.
Whatever sub-class type as well as Evaluator class type will be represented in BHoM via enumeration classes assigned to the RevitFilterRule and RevitFilterValueRule
as attributes.

GCRA101 committed:
Delete FilterRuleType Enum

Instead of using a single class FilterRule with an Enumeration encapsulated to represent all the different Revit API Filter Rules,
we go for creating multiple classes and sub-classes in BHoM as well, thus trying to match as much as possible the classes structure
that is present in the API.

GCRA101 committed:
Delete FilterRuleType Enum

Instead of using a single class FilterRule with an Enumeration encapsulated to represent all the different Revit API Filter Rules,
we go for creating multiple classes and sub-classes in BHoM as well, thus trying to match as much as possible the classes structure
that is present in the API.

GCRA101 committed:
Create FilterRule classes mirroring Revit API classes

GCRA101 committed:
Create FilterRule classes mirroring Revit API classes

GCRA101 committed:
Move FillPattern and LinePattern to the Enums Namespace

GCRA101 committed:
Move FillPattern and LinePattern to the Enums Namespace

GCRA101 committed:
Modify BHoM Enum TextComparisonType

Add NOT_START_WITH and NOT_ENDS_WITH enum values to the enumeration.
These are required to implement the View Filtering in BHoM.

GCRA101 committed:
Modify BHoM Enum TextComparisonType

Add NOT_START_WITH and NOT_ENDS_WITH enum values to the enumeration.
These are required to implement the View Filtering in BHoM.

GCRA101 committed:
Add LevelComparisonType Enumeration

Useful to create LevelFilterRule objects

GCRA101 committed:
Add LevelComparisonType Enumeration

Useful to create LevelFilterRule objects

GCRA101 committed:
Add conversion methods for FilterMaterialRule and FilterLevelRule in Convert.ToRevit Namespace

GCRA101 committed:
Add conversion methods for FilterMaterialRule and FilterLevelRule in Convert.ToRevit Namespace

GCRA101 committed:
Update conversion of BHoM ViewFilters to Revit Parameter Filter Elements

GCRA101 committed:
Update conversion of BHoM ViewFilters to Revit Parameter Filter Elements

GCRA101 committed:
Add enum values to TextComparisonType Enumeration

Greater, GreaterOrEqual, Less, LessOrEqual are added to reflect the types and names of
the subclasses of the Revit API class FilterStringRuleEvaluator

GCRA101 committed:
Add enum values to TextComparisonType Enumeration

Greater, GreaterOrEqual, Less, LessOrEqual are added to reflect the types and names of
the subclasses of the Revit API class FilterStringRuleEvaluator

GCRA101 committed:
Update ToRevit ParameterFilterElement with additional TextComparisonType Enum Values

Greater, GreaterOrEqual, Less, LessOrEqual

GCRA101 committed:
Update ToRevit ParameterFilterElement with additional TextComparisonType Enum Values

Greater, GreaterOrEqual, Less, LessOrEqual

GCRA101 committed:
Add adapter code for FilterRules in ViewFilterFromRevit Method in FromRevit Namespace

GCRA101 committed:
Add adapter code for FilterRules in ViewFilterFromRevit Method in FromRevit Namespace

GCRA101 committed:
Add FilterElementIdRule to the Revit BHoM Objects

GCRA101 committed:
Add FilterElementIdRule to the Revit BHoM Objects

GCRA101 committed:
Add handling of FilterElementIdRule in conversion of ViewFilter From Revit

GCRA101 committed:
Add handling of FilterElementIdRule in conversion of ViewFilter From Revit

GCRA101 committed:
Initialize FilterRule Classes in FromRevit and ToRevit Namespaces

Instead of keeping the conversion of FilterRule objects from/to Revit/BHoM within the ParameterFilter/ViewFilter
conversion classes only, we split the code in each different FilterRule Class keeping the code more modular and reusable.

GCRA101 committed:
Initialize FilterRule Classes in FromRevit and ToRevit Namespaces

Instead of keeping the conversion of FilterRule objects from/to Revit/BHoM within the ParameterFilter/ViewFilter
conversion classes only, we split the code in each different FilterRule Class keeping the code more modular and reusable.

GCRA101 committed:
Delete all FilterRule classes created in the Convert.FromRevit/ToRevit namespaces.

Any class that can be stored in the Convert.FromRevit/ToRevit namespace, has to be
convertible into Revit.DB.Element Class. Since FilterRule classes are not convertible into
Revit.DB.Elements, they cannot be present in the Convert.FromRevit/ToRevit.

GCRA101 committed:
Delete all FilterRule classes created in the Convert.FromRevit/ToRevit namespaces.

Any class that can be stored in the Convert.FromRevit/ToRevit namespace, has to be
convertible into Revit.DB.Element Class. Since FilterRule classes are not convertible into
Revit.DB.Elements, they cannot be present in the Convert.FromRevit/ToRevit.

GCRA101 committed:
Reinstate the method filterCategoryRuleToRevit in the Convert.ToRevit.ParameterFilterElement Class.

Deleted previously by mistake.

GCRA101 committed:
Reinstate the method filterCategoryRuleToRevit in the Convert.ToRevit.ParameterFilterElement Class.

Deleted previously by mistake.

GCRA101 committed:
Add method to convert a Revit ParameterFilterElement into a BHoM ViewFilter.

GCRA101 committed:
Add method to convert a Revit ParameterFilterElement into a BHoM ViewFilter.

GCRA101 committed:
Incorporate minor fixes to references between classes

GCRA101 committed:
Incorporate minor fixes to references between classes

GCRA101 committed:
Add ParameterValuePresenceRule

Add it as BHoM Class and implement code for conversion to/from Revit

GCRA101 committed:
Add ParameterValuePresenceRule

Add it as BHoM Class and implement code for conversion to/from Revit

GCRA101 committed:
Move all BHoM Filter Rules from oM.Elements to oM.FilterRules namespace.

In the oM.Elements namespace we can store only the BHoM object that have a direct corresponding object in Revit.

GCRA101 committed:
Move all BHoM Filter Rules from oM.Elements to oM.FilterRules namespace.

In the oM.Elements namespace we can store only the BHoM object that have a direct corresponding object in Revit.

GCRA101 committed:
Refactor the partial class FromRevit.ViewFilter

GCRA101 committed:
Refactor the partial class FromRevit.ViewFilter

GCRA101 committed:
Fix code for conversion of Revit Views into BHoM Views

GCRA101 committed:
Fix code for conversion of Revit Views into BHoM Views

GCRA101 committed:
Make explicit distinction between Revit and BHoM View Class instances

GCRA101 committed:
Make explicit distinction between Revit and BHoM View Class instances

GCRA101 committed:
Add conversion of BHoM FilterRule to Revit FilterRule

GCRA101 committed:
Add conversion of BHoM FilterRule to Revit FilterRule

GCRA101 committed:
Fix bug with extraction of revit Id of selected parameter

GCRA101 committed:
Fix bug with extraction of revit Id of selected parameter

GCRA101 committed:
Fix code for conversion of filterRules from BHoM to Revit_Core_Engine/Convert/Revit/ToRevit/ParameterFilterElement

Adapt code to work better with the restraints of the Revit API Rule Factory Classes

GCRA101 committed:
Fix code for conversion of filterRules from BHoM to Revit_Core_Engine/Convert/Revit/ToRevit/ParameterFilterElement

Adapt code to work better with the restraints of the Revit API Rule Factory Classes

GCRA101 committed:
Add call to method for convertion of FilterLevelRule to Revit.

GCRA101 committed:
Add call to method for convertion of FilterLevelRule to Revit.

GCRA101 committed:
Fix bug in creation of Revit Element Filter from BHoM FilterRules

GCRA101 committed:
Fix bug in creation of Revit Element Filter from BHoM FilterRules

GCRA101 committed:
Add changes to Level and Material FilterRule Conversion

GCRA101 committed:
Add changes to Level and Material FilterRule Conversion

GCRA101 committed:
Fix bug with Push of View Filters based on Levels

GCRA101 committed:
Fix bug with Push of View Filters based on Levels

GCRA101 committed:
Fix bug dispatching different types of FilterRule

GCRA101 committed:
Fix bug dispatching different types of FilterRule

GCRA101 committed:
Turn FilterNumericValueRule into an abstract class

In this way its corresponding Grasshopper Component will disappear, thus forcing the user
to use only its specific sub-components: DoubleRule, IntegerRule and ElementIdRule.

GCRA101 committed:
Turn FilterNumericValueRule into an abstract class

In this way its corresponding Grasshopper Component will disappear, thus forcing the user
to use only its specific sub-components: DoubleRule, IntegerRule and ElementIdRule.

GCRA101 committed:
Fix bug related to incorrect extraction of Categories from the model.

Rather than using a FilteredElementCollector on which to call the .OfClass() method,
better to get all the Category objects getting the Categories data structure from the Revit model and
then iterate on its elements to extract all the categories and transfer them into a list.

GCRA101 committed:
Fix bug related to incorrect extraction of Categories from the model.

Rather than using a FilteredElementCollector on which to call the .OfClass() method,
better to get all the Category objects getting the Categories data structure from the Revit model and
then iterate on its elements to extract all the categories and transfer them into a list.

GCRA101 committed:
Add feature to create ViewFilters with No Rules.

GCRA101 committed:
Add feature to create ViewFilters with No Rules.

GCRA101 committed:
Fix code for extraction of categories and filter rules from Revit when pulling

GCRA101 committed:
Fix code for extraction of categories and filter rules from Revit when pulling

GCRA101 committed:
Add GetParameterById Utility function and minor updates to parameter collection streams

GCRA101 committed:
Add GetParameterById Utility function and minor updates to parameter collection streams

GCRA101 committed:
Add use of utility function GetParameterById where required to get parameter objects

GCRA101 committed:
Add use of utility function GetParameterById where required to get parameter objects

GCRA101 committed:
Add conversion of double internal units when pulling filter parameter values

GCRA101 committed:
Add conversion of double internal units when pulling filter parameter values

GCRA101 committed:
Add Update Method for Autodesk Revit View3D Class

The View3D Class is a subclass of the View Class. To let the BHoM run the Update method on the View3D class, we need to specify an overloaded version of the Update Method taking a View3D class instance rather than a View class instance as input.

GCRA101 committed:
Add Update Method for Autodesk Revit View3D Class

The View3D Class is a subclass of the View Class. To let the BHoM run the Update method on the View3D class, we need to specify an overloaded version of the Update Method taking a View3D class instance rather than a View class instance as input.

GCRA101 committed:
Turn Update Methods for View and View3D into a single method using Generic Type

Since View3D is a subclass of View in the Revit API, we can make the Update method work on a generic type T, specifying that it must be extending the revit api superclass View.

GCRA101 committed:
Turn Update Methods for View and View3D into a single method using Generic Type

Since View3D is a subclass of View in the Revit API, we can make the Update method work on a generic type T, specifying that it must be extending the revit api superclass View.

GCRA101 committed:
Fix Fill and Line Pattern Name search turning letter case to Upper

GCRA101 committed:
Fix Fill and Line Pattern Name search turning letter case to Upper

GCRA101 committed:
Fix bug text comparison between Cut and Surface Patterns

GCRA101 committed:
Fix bug text comparison between Cut and Surface Patterns

pawelbaran commented on Revit_Core_Engine/Convert/Revit/FromRevit/View.cs:
```diff
@@ -0,0 +1,135 @@
+                return view;
+
+            /*1. Transfer ViewFilters and corresponding OverrideGraphicSettings into ViewFilterWithOverrides Objects - via STREAMS */
+            List<ViewFilterWithOverrides> filtersWithOverrides = revitView.GetFilters().ToDictionary<ElementId, ViewFilter, OverrideGraphicSettings>
```
Some view types (e.g. project browser) do not support overrides, causing the code to crash in case of a simple pull:
![image](https://github.com/user-attachments/assets/037a53c8-d9da-4a17-a952-fd00cdb2d153)

GCRA101 replied:
@pawelbaran, thanks for flagging this.
I've added an  if statement on the boolean parameter "AreGraphicsOverridesAllowed()" of the Autodesk View Class.
Whenever the parameter is true, the method GetFilters() gets run and no error messages get triggered.
Whenever the parameter is false, the corresponding BHoM view gets created without any filters assigned to it.

![image](https://github.com/user-attachments/assets/0c105807-89f0-41e9-84ec-8c3fe256f4de)

pawelbaran commented on Revit_Core_Engine/Convert/Revit/FromRevit/View.cs:
```diff
@@ -0,0 +1,135 @@
+            OverrideGraphicSettings overrideGraphicsSettings = new OverrideGraphicSettings();
+
+            // Convert COLORS 
+            overrideGraphicsSettings.LineColor = System.Drawing.Color.FromArgb(revitOverrides.CutLineColor.Red, revitOverrides.CutLineColor.Green, revitOverrides.CutLineColor.Blue);
```
You need to check if `revitOverrides.CutLineColor.IsValid` first, otherwise it crashes in case of no color override applied

pawelbaran replied:
As discussed live, need to avoid mixing dash & dot, so if (contains "Dash" but not "Dot") etc.

GCRA101 replied:
@pawelbaran , I've added an if statement checking validity of attribute

pawelbaran commented on Revit_Core_Engine/Convert/Revit/FromRevit/View.cs:
```diff
@@ -0,0 +1,135 @@
+
+
+            // Convert CUT PATTERNS
+            String cutPatternName = element.Document.GetElement(revitOverrides.CutBackgroundPatternId).Name;
```
You need to address the case where `CutLinePatternId == -1` (and all other Ids pointing at patterns, linestyles etc.)

pawelbaran reviewed (changes requested):
I went through the code and did some testing. Main takeaways:
- the general setup is mainly correct, which is great 🔥 
- most of the code comments are focused on bugs or code structure, so should be possible to address them one by one
- I had hard times to follow your test files for two reasons:
    - test scripts contain nodes from external packages, would be great to remove these (think of opening these files in two years from now - high chance the packages will be outdated by then)
    - the logic is quite packed, probably we should go through them together
- instead, I built my own testset and discovered quite a few bugs, see [here](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?FolderCTID=0x0120008122C8891F89054B8ACED0196C70DFC4&id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F02%5FPull%20Request%2FBHoM%2FRevit%5FToolkit%2F%231502%2DAddViewFilters%2FPB)
- it would be good to clean up usings and remove redundant qualification of types by namespace `Autodesk.Revit.DB` (which you even added in files that you have not touched otherwise)

Finally, we will need to combine `View` and `ViewPlan` classes in Revit_oM, but that one can be treated as a separate exercise to be targeted once we get the concept of filters sorted.

pawelbaran commented on Revit_Core_Engine/Convert/Revit/FromRevit/View.cs:
```diff
@@ -0,0 +1,135 @@
+            // Convert LINE PATTERNS
+            String linePatternName = element.Document.GetElement(revitOverrides.CutLinePatternId).Name;
+
+            if (linePatternName.ToUpper().Contains("DASH")) { overrideGraphicsSettings.LinePattern = oM.Revit.Enums.LinePattern.DASH; }
```
As discussed live, need to avoid mixing dash & dot, so if (contains "Dash" but not "Dot") etc.

GCRA101 replied:
@pawelbaran, added new more specific enums for line patterns to make it easier the match between BHoM and Revit Line Pattern objects.
![image](https://github.com/user-attachments/assets/b69ce7e9-51ce-4c4c-847e-fc70d57331c0)

pawelbaran commented on Revit_Core_Engine/Convert/Revit/FromRevit/ViewFilter.cs:
```diff
@@ -0,0 +1,523 @@
+                NumberComparisonType bhomNumericEvaluator = 0;
+
+                // FILTER STRING RULE
+                if (revitRule.GetType()== typeof(Autodesk.Revit.DB.FilterStringRule))
+                { return (oM.Revit.FilterRules.FilterRule)FilterStringRuleFromRevit(revitViewFilter, revitRule, bhomTextEvaluator); }
+                // FILTER DOUBLE RULE
+                else if (revitRule.GetType()==typeof(Autodesk.Revit.DB.FilterDoubleRule))
+                { return (oM.Revit.FilterRules.FilterRule)FilterDoubleRuleFromRevit(revitViewFilter, revitRule, bhomNumericEvaluator); }
+                // FILTER INTEGER RULE
+                else if (revitRule.GetType() == typeof(Autodesk.Revit.DB.FilterIntegerRule))
+                { return (oM.Revit.FilterRules.FilterRule)FilterIntegerRuleFromRevit(revitViewFilter, revitRule, bhomNumericEvaluator); }
+                // FILTER ELEMENTID RULE
+                else if (revitRule.GetType() == typeof(Autodesk.Revit.DB.FilterElementIdRule))
+                { return (oM.Revit.FilterRules.FilterRule)FilterElementIdRuleFromRevit(revitViewFilter, revitRule, bhomNumericEvaluator); }
+                // FILTER CATEGORY RULE
+                else if (revitRule.GetType() == typeof(Autodesk.Revit.DB.FilterCategoryRule))
+                { return (oM.Revit.FilterRules.FilterRule)FilterCategoryRuleFromRevit(revitViewFilter, revitRule); }
+                // FILTER PARAMETER VALUE PRESENCE RULE
+                else if (revitRule.GetType() == typeof(Autodesk.Revit.DB.ParameterValuePresenceRule))
+                { return (oM.Revit.FilterRules.FilterRule)ParameterValuePresenceRuleFromRevit(revitViewFilter, revitRule); }
+                // FILTER INVERSE RULE
+                else if (revitRule.GetType() == typeof(Autodesk.Revit.DB.FilterInverseRule))
+                { return FilterInverseRuleFromRevit(revitViewFilter, revitRule, bhomTextEvaluator, bhomNumericEvaluator); }
+                return null;
```
...then you need to add overloads of `FilterRuleFromRevit` method for each of the subtypes to dynamically bind to

GCRA101 replied:
@pawelbaran, thanks a lot for this great tip. A true jewel ;D
I've updated the code in line with it using the dynamic casting .

Now that long series of if statements is condensed in 2 lines of code ;)

![image](https://github.com/user-attachments/assets/3bedbd6c-ad20-45dd-8c55-aa2f45745008)

pawelbaran commented on Revit_Core_Engine/Modify/Update.cs:
```diff
@@ -243,6 +249,91 @@ public static bool Update(this AssemblyInstance element, Assembly bHoMObject, Re
+        public static bool Update(this ParameterFilterElement element, BH.oM.Adapters.Revit.Elements.ViewFilter bHoMObject, RevitSettings settings, bool setLocationOnUpdate)
+        {
+            // 1. Collect the ElementIds of the filter's categories - via Streams
+            List<ElementId> builtInCategoryIds= bHoMObject.Categories.Select(catName => { BuiltInCategory builtInCat;
```
`catName`, checked it myself ultimately

pawelbaran commented on Revit_oM/Elements/ViewFilter.cs:
```diff
@@ -0,0 +1,17 @@
+{
+    public class ViewFilter : BHoMObject
+    {
+        public virtual List<string> Categories {  get; set; }
```
Actually, worth checking if it would not make sense to use https://github.com/BHoM/Revit_Toolkit/blob/develop/Revit_oM/Enums/Category.cs

GCRA101 replied:
@pawelbaran , thanks for the comment and for the tip.
Replaced the string list with a List of BHoM Category Enums.

pawelbaran commented on Revit_oM/Enums/FillPattern.cs:
```diff
@@ -0,0 +1,20 @@
+
+namespace BH.oM.Revit.Enums
+{
+    public enum FillPattern
```
`DiagonalCrosshatch` for example

GCRA101 replied:
@pawelbaran, enums updated using PascalCasing

GCRA101 commented on Revit_Core_Engine/Convert/Revit/FromRevit/ViewFilter.cs:
```diff
@@ -0,0 +1,523 @@
+                            .Select(filter => ((ElementParameterFilter)filter).GetRules())
+                            .SelectMany(list => list)
+                            .ToDictionary(rule => rule, rule => (GetParameterById(revitViewFilter.Document, rule.GetRuleParameter()).Definition.Name.ToString()));
+                            //.ToList<string>();
```
@pawelbaran thanks for the comment.
commented out code lines have been removed (see latest commit)

GCRA101 commented on Revit_Core_Engine/Convert/Revit/FromRevit/ViewFilter.cs:
```diff
@@ -0,0 +1,523 @@
+                    Dictionary<FilterRule, string> filterAndParNames = ((ElementLogicalFilter)revitViewFilter.GetElementFilter()).GetFilters()
+                            .Select(filter => ((ElementParameterFilter)filter).GetRules())
+                            .SelectMany(list => list)
+                            .ToDictionary(rule => rule, rule => (GetParameterById(revitViewFilter.Document, rule.GetRuleParameter()).Definition.Name.ToString()));
```
@pawelbaran, change incorporated in latest commit.

GCRA101 commented on Revit_oM/Enums/LinePattern.cs:
```diff
@@ -0,0 +1,16 @@
+{
+    public enum LinePattern
+    {
+        DASH,
+        DOT,
+        HIDDEN,
+        SOLID
```
@pawelbaran, enum updated using PascalCasing

GCRA101 commented on Revit_Core_Engine/Modify/Update.cs:
```diff
@@ -323,6 +414,9 @@ private static void ErrorOnCategoryChange(this FamilyInstance element, IInstance
         }
 
         /***************************************************/
+
+
+      
```
empty lines removed

GCRA101 commented on Revit_oM/FilterRules/FilterStringRule.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    public class FilterStringRule : FilterValueRule
+    {
+        public virtual TextComparisonType Evaluator { get; set; }
```
@pawelbaran , Evaluator has been replaced with ComparisonType as suggested.

GCRA101 commented on Revit_oM/FilterRules/ParameterValuePresenceRule.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace BH.oM.Revit.FilterRules
+{
+    public class ParameterValuePresenceRule : FilterValueRule
```
New abstract class FilterParameterRule has been created inheriting from FIlterRule. ParameterValuePresenceRule class now inheriting from FilterParameterRule.

GCRA101 commented on Revit_Core_Engine/Modify/Update.cs:
```diff
@@ -243,6 +249,91 @@ public static bool Update(this AssemblyInstance element, Assembly bHoMObject, Re
+        [Input("settings", "Revit adapter settings to be used while performing the action.")]
+        [Input("setLocationOnUpdate", "Revit View instance does not have location property, therefore this parameter is irrelevant.")]
+        [Output("success", "True if the Revit View instance has been updated successfully based on the input BHoM View.")]
+        public static bool Update<T>(this T element, BH.oM.Adapters.Revit.Elements.View bHoMObject, RevitSettings settings, bool setLocationOnUpdate) where T:Autodesk.Revit.DB.View
```
@pawelbaran, not sure why but as soon as "Autodesk.Revit.DB" gets removed, VS doesn't manage to figure out whether the View is an Autodesk or a BHoM object. I'll leave it as it is for now.

pawelbaran replied:
Ha, you were right from the very beginning! What I was missing was the fact that this PR is adding `View` class to BHoM, which now starts to cause ambiguity against the `Autodesk.Revit.DB` type. So, we obviously need to specify full namespace here. 👍 Apologies for the confusion

GCRA101 committed:
Incorporate PR comments from Pawel Baran - Part01

GCRA101 committed:
Incorporate PR comments from Pawel Baran - Part01

GCRA101 committed:
Add new abstract class FilterParameterRule

Using this abstract class it's possible to decouple the ParameterValuePresence Class from the FilterValueRule class.

GCRA101 committed:
Add new abstract class FilterParameterRule

Using this abstract class it's possible to decouple the ParameterValuePresence Class from the FilterValueRule class.

GCRA101 committed:
Incorporate PR comments from Pawel Baran - Part02

GCRA101 committed:
Incorporate PR comments from Pawel Baran - Part02

GCRA101 commented on Revit_Core_Engine/Convert/Revit/FromRevit/ViewFilter.cs:
```diff
@@ -0,0 +1,523 @@
+        {
+            List<oM.Revit.FilterRules.FilterRule> bhomFilterRules = revitFilterRules.Select(revitRule =>
+            {
+                TextComparisonType bhomTextEvaluator = 0;
+                NumberComparisonType bhomNumericEvaluator = 0;
```
@pawelbaran, thanks for the comment.
Removed these 2 variables from the outer method and declared them directly in the inner methods that are using them directly.

GCRA101 commented on Revit_Core_Engine/Modify/Update.cs:
```diff
@@ -243,6 +249,91 @@ public static bool Update(this AssemblyInstance element, Assembly bHoMObject, Re
+                    FillPatternElement revitCutPattern = new FilteredElementCollector(element.Document)
+                        .OfClass(typeof(FillPatternElement))
+                        .Cast<FillPatternElement>()
+                        .FirstOrDefault(pattern => pattern.Name.ToUpper().Contains(kvp.Value.CutPattern.ToString().Replace("_"," ").ToUpper()));
```
Added DashDot, DashDotDot and other Line Patterns to BHoM Enum. Updated using Pascal Casing rather than Upper Case and Underscores.

GCRA101 commented on Revit_Core_Engine/Convert/Revit/FromRevit/ViewFilter.cs:
```diff
@@ -0,0 +1,523 @@
+            FilterStringRuleEvaluator stringEvaluator = revitFilterStringRule.GetEvaluator();
+
+            // Convert the REVIT FilterStringEvaluator type into the BHOM TextComparisonType Enum 
+            switch (stringEvaluator.GetType().ToString())
+            {
+                case "Autodesk.Revit.DB.FilterStringBeginsWith":
+                    bhomTextEvaluator = TextComparisonType.StartsWith;
+                    break;
```
@pawelbaran thanks for this. I've updated the code in line with the above. That's great ;)

GCRA101 commented on Revit_Core_Engine/Convert/Revit/FromRevit/ViewFilter.cs:
```diff
@@ -0,0 +1,523 @@
+                case ("Autodesk.Revit.DB.FilterStringRule"):
+                    {
+                        switch (((Autodesk.Revit.DB.FilterStringRule)innerRule).GetEvaluator().GetType().ToString())
+                        {
```
Done as per your suggestion

GCRA101 commented on Revit_Core_Engine/Convert/Revit/FromRevit/ViewFilter.cs:
```diff
@@ -0,0 +1,523 @@
+            FilterNumericRuleEvaluator numericEvaluator = revitFilterDoubleRule.GetEvaluator();
+
+            // Convert the REVIT FilterNumericEvaluator type into the BHOM NumberComparisonType Enum 
+            switch (numericEvaluator.GetType().ToString())
```
@pawelbaran, 
thanks for the comment and the suggestion. 
Created new private utility methods encapsulating the switch/case code. ;)
Much cleaner code now ;)

GCRA101 committed:
Incorporate comments from Pawel Baran - Part03

GCRA101 committed:
Incorporate comments from Pawel Baran - Part03

GCRA101 commented on Revit_Core_Engine/Convert/Revit/FromRevit/ViewFilter.cs:
```diff
@@ -0,0 +1,523 @@
+                // FILTER PARAMETER VALUE PRESENCE RULE
+                else if (revitRule.GetType() == typeof(Autodesk.Revit.DB.ParameterValuePresenceRule))
+                { return (oM.Revit.FilterRules.FilterRule)ParameterValuePresenceRuleFromRevit(revitViewFilter, revitRule); }
+                // FILTER INVERSE RULE
```
@pawelbaran,
the HasValueFilterRule and HasNoValueFilterRule are managedby exclusion using the code line bhomParamValuePresenceRule.IsPresent = (revitRule.GetType() == typeof(HasValueFilterRule)) ? true : false;

In the Revit API HasValueFilterRule and HasNoValueFilterRule inherit both from ParameterValuePresenceRule. 
![image](https://github.com/user-attachments/assets/d185c827-5634-44d6-bdf5-b8418e45ee54)

The two Revit API classes above are managed by our BHoM class ParameterValuePresenceRule via the boolean attribute isPresent.
![image](https://github.com/user-attachments/assets/85cd6498-c15a-4bc6-8090-a7da43aac0bf)

I would say that it's doing what we are expecting but let me know in case I'm missing something ;)

pawelbaran replied:
What I think you are missing here is the following line:
```C#
else if (revitRule.GetType() == typeof(Autodesk.Revit.DB.ParameterValuePresenceRule))
```

It checks whether the rule type is exactly `ParameterValuePresenceRule`, ignoring inheritance, so it will only catch this exact type. Meanwhile, its children (`HasNoValueFilterRule`, `HasValueFilterRule`) will not be caught. If you want to capture all 3 types at once, then instead of `GetType() ==` you should rather use `is` statement.

GCRA101 commented on Revit_Core_Engine/Convert/Revit/FromRevit/DraftingInstance.cs:
```diff
@@ -53,7 +53,7 @@ public static DraftingInstance DraftingInstanceFromRevit(this FilledRegion fille
             if (draftingInstance != null)
                 return draftingInstance;
 
-            View view = filledRegion.Document.GetElement(filledRegion.OwnerViewId) as View;
```
@pawelbaran, 
as discussed via Teams, the requirement for presence of all namespaces list in name of the View is due to the fact that the View class is present in both Autodesk and BHoM.
All sorted ;)

GCRA101 commented on Revit_Core_Engine/Convert/Revit/FromRevit/IInstance.cs:
```diff
@@ -55,7 +55,7 @@ public static IInstance InstanceFromRevit(this CurveElement curveElement, RevitS
 
             if (curveElement.ViewSpecific)
             {
-                View view = curveElement.Document.GetElement(curveElement.OwnerViewId) as View;
```
@pawelbaran,
as discussed via Teams, the requirement for presence of all namespaces list in name of the View is due to the fact that the View class is present in both Autodesk and BHoM.
All sorted ;)

GCRA101 commented on Revit_Core_Engine/Convert/Revit/FromRevit/ViewFilter.cs:
```diff
@@ -0,0 +1,523 @@
+                    Dictionary<FilterRule, string> filterAndParNames = ((ElementLogicalFilter)revitViewFilter.GetElementFilter()).GetFilters()
+                            .Select(filter => ((ElementParameterFilter)filter).GetRules())
+                            .SelectMany(list => list)
+                            .ToDictionary(rule => rule, rule => (GetParameterById(revitViewFilter.Document, rule.GetRuleParameter()).Definition.Name.ToString()));
```
@pawelbaran,
incorporating previous comment, I've turned the dictionary into a List as the code doesn't actually need the parameterbyId. Therefore, no need to try to get the parameterrule out of the FilterRule.
The bug shouldn't come up anymore.

GCRA101 committed:
Incorporate PR Comments from Pawel Baran - Part04

GCRA101 committed:
Incorporate PR Comments from Pawel Baran - Part04

GCRA101 commented on Revit_Core_Engine/Convert/Revit/FromRevit/ViewFilter.cs:
```diff
@@ -0,0 +1,523 @@
+
+
+        private static oM.Revit.FilterRules.FilterStringRule FilterStringRuleFromRevit(this ParameterFilterElement revitViewFilter,
+            Autodesk.Revit.DB.FilterRule revitRule, TextComparisonType bhomTextEvaluator)
```
Comment incorporated

GCRA101 committed:
Incorporate PR comments from Pawel Baran - Part 05

GCRA101 committed:
Merge modifications

GCRA101 committed:
Fix merge conflict

GCRA101 committed:
Fix minor bugs with dynamic casting and conversion between Category Enum values and Revit Category Objects

pawelbaran commented on Revit_Core_Engine/Modify/Update.cs:
```diff
@@ -243,6 +249,91 @@ public static bool Update(this AssemblyInstance element, Assembly bHoMObject, Re
+        [Input("settings", "Revit adapter settings to be used while performing the action.")]
+        [Input("setLocationOnUpdate", "Revit View instance does not have location property, therefore this parameter is irrelevant.")]
+        [Output("success", "True if the Revit View instance has been updated successfully based on the input BHoM View.")]
+        public static bool Update<T>(this T element, BH.oM.Adapters.Revit.Elements.View bHoMObject, RevitSettings settings, bool setLocationOnUpdate) where T:View
```
I am still not sure why would we need a generic here, couldn't the signature looks as follows:
```suggestion
        public static bool Update(this View element, BH.oM.Adapters.Revit.Elements.View bHoMObject, RevitSettings settings, bool setLocationOnUpdate)
```

GCRA101 replied:
@pawelbaran, thanks for reflagging this.
I've just forgot it.
Now incorporated in the code.

pawelbaran commented on Revit_Core_Engine/Convert/Revit/FromRevit/ViewFilter.cs:
```diff
@@ -0,0 +1,486 @@
+            // 1. EXTRACT DATA from the REVIT FILTERRULE object
+
+            // Downcast Revit FilterRule obj to Revit FilterCategoryRule obj
+            Autodesk.Revit.DB.FilterCategoryRule revitFilterCategoryRule = (Autodesk.Revit.DB.FilterCategoryRule)revitRule;
```
No need to do this casting any more (refers to all private methods that do it in this file)

GCRA101 replied:
Removed all the casts in the code.

pawelbaran commented:
@BHoMBot check compliance

pawelbaran commented on Revit_Core_Engine/Convert/Revit/FromRevit/View.cs:
```diff
@@ -0,0 +1,164 @@
+
+        /***************************************************/
+
+        private static OverrideGraphicSettings overrideGraphicSettingsFromRevit(this View element, Autodesk.Revit.DB.OverrideGraphicSettings revitOverrides)
```
```suggestion
        private static OverrideGraphicSettings OverrideGraphicSettingsFromRevit(this View element, Autodesk.Revit.DB.OverrideGraphicSettings revitOverrides)
```

pawelbaran commented on Revit_Core_Engine/Convert/Revit/FromRevit/View.cs:
```diff
@@ -0,0 +1,164 @@
+            // Convert LINE PATTERNS
+            try 
+                { 
+                    String linePatternName = element.Document.GetElement(revitOverrides.CutLinePatternId).Name;
```
Instead of wrapping all of that in try/catch it should be enough to check if `revitOverrides.CutLinePatternId != -1`. Or am I missing a point?

GCRA101 replied:
@pawel, thanks for this. 
No specific reason to go for try/catch rather than if/then statement.
Updated the code in line with your comments.

pawelbaran replied:
FYI landing in `catch` is extremely slow - it triggers the runtime to do reporting, generate stack trace, dump memory etc. For that reason it should be last resort rather than an alternative to ifs 👍

GCRA101 committed:
Update name of method OverrideGraphicSettings#fromRevit based on Pascal Casing

Co-authored-by: Pawel Baran <pawel.baran@burohappold.com>

GCRA101 committed:
Remove FilterRule downcasting and use of generic T type in Update Method for Views

GCRA101 committed:
Replace Try-catch statements with If-then statments for controlling errors in PatternId assigment

GCRA101 commented:
@pawelbaran, thanks for your comments.
Updated code now in latest commit

pawelbaran commented:
Thanks for the activity @GCRA101 - can see your commits addressing my comments on the ongoing basis 😃 I think the code is OK now, I am slowly switching to testing. Found quite a few bugs so far:
- filters with `Workset` still throw error on pull
    ![image](https://github.com/user-attachments/assets/12af8a16-6616-4086-bc21-f1a0bd828251)
- same as above for has value and does not have value
- updating categories using Push/Update causes reset in the filter (which kind of makes sense - after modifying categories one would need to check if all parameters are available in each of them, so not surprised it breaks things, it almost feels like updating categories should not be allowed)
- changing any parameter in the filter resets all parameters in the filter

All above cases are covered in the test materials I updated [here](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?FolderCTID=0x0120008122C8891F89054B8ACED0196C70DFC4&id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F02%5FPull%20Request%2FBHoM%2FRevit%5FToolkit%2F%231502%2DAddViewFilters%2FPB). Please note these materials cover only the most basic cases, you'll need to test the code against real life models too. But this is the last step, let's focus on getting the basics sorted first 👍
